### DISCUSSION TITLE
Method for fast full screen updates from a buffer?

### DISCUSSION THREAD

**Discussion by vortigont** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#117] [Ideas]
I've got that idea of pipe-lining when I was playing with fillrate and implementing drawFast*methods. I found that function calls under esp32 is pretty heavy itself for CPU cycles, maybe it is related to instruction cache misses or whatever, not sure... But changing full row of pixels in a single tight loop is much faster than doing the same via drawpixel and external loop iterating the line. So let's take an example with some of the Aurora demo's - it works in a semi double-buffer mode. I mean it has it's own buffer in a CRGB vector for a single frame of effect and also DMA buffer for the matrix. First stage is calculating frame for effect over CRGB vector and second stage - copy vector buffer to the DMA buffer via drawPixel() using one pixel at a time. Now if we could use some kind of a drawFastHLine() but the one that accepts not just a single color for all pixels of a line but a reference to the vector buffer (offset and length) than the same approach could fill the entire row for the DMA buffer in a single loop. I expect it to be about 3 times faster than per-pixel approach with an external loop, maybe even more if being able to update two rows at once, top and bottom half of the same 16 bit word of the DMA buff.
But than comes but's and if's - it is pretty easy to implement for a single panel, but for virtual panels it's not that easy to provide a buffer vector matching a full virtual row covering the entire display. So, that's why I'm still thing about the way this could (or could not) be done in a nice way.
Any ideas?

**Comment by mrcodetastic** [comment]
I love your enthusiasm @vortigont, but all I see is pain.

My argument is if there's a desperate need to extract a last drop of GFX performance out of the ESP32, and essentially draw bare to the DMA memory without any of the adafruit abstraction, then they ought to fork the library and implement their graphics / sketch logic in the library itself lol and write directly to the malloc created for each row. 

The point of this library was to leverage the DMA capabilities, and abstract the pain away of doing so.

**Reply by vortigont** [reply]
Agreed on leaving all the pain aside :) That's why I've done only simple lines/rects. But the idea to have same low-level implementation of updating DMA buffer in batches is too sweet to just throw it away :) Bit logic for DMA buff is quite complex to deep dive into it each time you need to cut some corners and squeeze more FPS from a single ESP32 :) As a general idea I would like to have a fastHLine-like method to render some arbitrary RGB vector. I believe it must be as simple and fast as possible, no coordinates transformations or H-to-V transpose, viewports, etc... just a row number and length. The idea is provide primitive to directly map RGB vector to the DMA buffer of the same pixel size,
All other overlays must live somewhere else - child classes, virtual panels, whatever... That would leave the freedom to concentrate the efforts on transpose logic operating with 8 bit RGB buffs only and keep your head (and hands) out of I2S bitlogic.
Well, anyway, do not have free time for experiments for now, unfortunately. But always open for discussions :)

**Comment by mrcodetastic** [comment]
If one wants to implement a bare-metal graphics then starting with this sketch would be the way to go: https://www.esp32.com/viewtopic.php?f=17&t=3188

Funnily enough, ESP_Sprite's example code was the basis for this library originally... but subsequently built atop to make it Adafruit compatible (+ your excellent additions @vortigont) etc.

Edit: Looks like somebody made a full-screen style buffer library, but obviously you lose the ease associated with this library: https://github.com/phkehl/esp32-leddisplay

**Reply by vortigont** [reply]
that esp32 forum post is nice place to start for the one who want to write... an i2s library? Nah... Sometimes I want to stay within good old RGB bounds :)
BTW I've tried raspberry with the same RGB panels I have. Can't say it does much more better job than this lib :) It definitely has  more RAM and can do parallel output, but the refresh rate for a single channel is pretty much the same. And it eats one CPU core completely. So I must admit that I2S DMA on ESP32 is not that bad at all!

**Reply by mrcodetastic** [reply]
I think newer revisions of the ESP32 coming in 2022+ (RISC-V version?) may support DMA from PSRAM/external SRAM etc. So that will be interesting to adjust this library for. i.e. FHD framebuffer display powered by an ESP32 might be possible lol.
